File: csharp/AstroOrbitData.cs
using System;
using UnityEngine;

// Token: 0x0200019A RID: 410
public class AstroOrbitData
{
	// Token: 0x0600121B RID: 4635 RVA: 0x00138B00 File Offset: 0x00136D00
	public void PredictPose(long time, VectorLF3 center, ref AstroData astroData)
	{
		double num = 40000.0 * (double)this.orbitRadius;
		double num2 = (double)time / (this.orbitalPeriod * 60.0) + (double)this.orbitPhase / 360.0;
		int num3 = (int)(num2 + 0.1);
		num2 -= (double)num3;
		this.runtimeOrbitPhase = (float)num2 * 360f;
		num2 *= 6.283185307179586;
		astroData.uPos = Maths.QRotateLF(this.orbitRotation, new VectorLF3((float)Math.Cos(num2), 0f, (float)Math.Sin(num2)));
		astroData.uRot = Quaternion.LookRotation(this.orbitNormal, astroData.uPos);
		astroData.uPos.x = astroData.uPos.x * num + center.x;
		astroData.uPos.y = astroData.uPos.y * num + center.y;
		astroData.uPos.z = astroData.uPos.z * num + center.z;
		double num4 = (double)(time + 1L) / (this.orbitalPeriod * 60.0) + (double)this.orbitPhase / 360.0;
		int num5 = (int)(num4 + 0.1);
		num4 -= (double)num5;
		num4 *= 6.283185307179586;
		astroData.uPosNext = Maths.QRotateLF(this.orbitRotation, new VectorLF3((float)Math.Cos(num4), 0f, (float)Math.Sin(num4)));
		astroData.uRotNext = Quaternion.LookRotation(this.orbitNormal, astroData.uPosNext);
		astroData.uPosNext.x = astroData.uPosNext.x * num + center.x;
		astroData.uPosNext.y = astroData.uPosNext.y * num + center.y;
		astroData.uPosNext.z = astroData.uPosNext.z * num + center.z;
	}

	// Token: 0x0600121C RID: 4636 RVA: 0x00138D08 File Offset: 0x00136F08
	public VectorLF3 GetVelocityAtPoint(VectorLF3 center, VectorLF3 upos)
	{
		VectorLF3 vectorLF;
		vectorLF.x = upos.x - center.x;
		vectorLF.y = upos.y - center.y;
		vectorLF.z = upos.z - center.z;
		return Maths.QRotateLF(Quaternion.AngleAxis(-360f / (float)(this.orbitalPeriod * 60.0), this.orbitNormal), vectorLF) - vectorLF;
	}

	// Token: 0x0600121D RID: 4637 RVA: 0x00138D84 File Offset: 0x00136F84
	public float GetEstimatePointOffset(float eta)
	{
		double num = 0.017453292519943295 * (double)eta * 360.0 / this.orbitalPeriod;
		return (float)((double)this.orbitRadius * 40000.0 * num * num * 0.5);
	}

	// Token: 0x040015AC RID: 5548
	public float orbitRadius = 1f;

	// Token: 0x040015AD RID: 5549
	public float orbitInclination;

	// Token: 0x040015AE RID: 5550
	public float orbitLongitude;

	// Token: 0x040015AF RID: 5551
	public double orbitalPeriod = 3600.0;

	// Token: 0x040015B0 RID: 5552
	public float orbitPhase;

	// Token: 0x040015B1 RID: 5553
	public float runtimeOrbitPhase;

	// Token: 0x040015B2 RID: 5554
	public Quaternion orbitRotation;

	// Token: 0x040015B3 RID: 5555
	public VectorLF3 orbitNormal;
}

File: src/data_struct/astro_orbit_data.rs
use crate::data_struct::astro_data::AstroData;
use crate::data_struct::consts::PI;
use crate::data_struct::vectors::{Quaternion, VectorLF3, LocalQuaternion};

pub struct AstroOrbitData {
    orbit_radius: f32,
    orbit_inclination: f32,
    orbit_longitude: f32,
    orbital_period: f64,
    orbit_phase: f32,
    runtime_orbit_phase: f32,
    orbit_rotation: Quaternion,
    orbit_normal: VectorLF3,
}

impl AstroOrbitData {
    pub fn new() -> Self {
        Self {
            orbit_radius: 0.0,
            orbit_inclination: 0.0,
            orbit_longitude: 0.0,
            orbital_period: 0.0,
            orbit_phase: 0.0,
            runtime_orbit_phase: 0.0,
            orbit_rotation: Quaternion::zeros(),
            orbit_normal: VectorLF3::zeros(),
        }
    }

    pub fn predict_pose(&mut self, time: i64, center: VectorLF3, astro_data: &mut AstroData) {
        let num = 40000.0 * self.orbit_radius as f64;
        let mut num2: f64 =
            time as f64 / (self.orbital_period as f64 * 60.0) + self.orbit_phase as f64 / 360.0;
        let num3 = (num2 + 0.1).floor();
        num2 -= num3;
        self.runtime_orbit_phase = (num2 * 360.0) as f32;
        num2 *= 2.0 * PI;
        astro_data.u_pos =
            self.orbit_rotation
                .rotation_lf(VectorLF3::new(num2.cos(), 0.0, num2.sin()));
        astro_data.u_rot = Quaternion::look_rotation(self.orbit_normal, astro_data.u_pos);
        astro_data.u_pos = astro_data.u_pos * num + center;
        let mut num4 =
            (time as f64 + 1.0) / (self.orbital_period * 60.0) + self.orbit_phase as f64 / 360.0;
        let num5 = (num4 + 0.1).floor();
        num4 -= num5;
        // *= 6.283185307179586
        num4 *= 2.0 * PI;
        astro_data.u_pos_next =
            self.orbit_rotation
                .rotation_lf(VectorLF3::new(num4.cos(), 0.0, num4.sin()));
        astro_data.u_rot_next = Quaternion::look_rotation(self.orbit_normal, astro_data.u_pos_next);
        astro_data.u_pos_next = astro_data.u_pos_next * num + center;
    }

    pub fn get_velocity_at_point(&self, center: VectorLF3, u_pos: VectorLF3) -> VectorLF3 {
        let mut vector_lf = u_pos - center;
        self.orbit_rotation.rotation_lf(Quaternion::angle_axis(
            -360.0 / (self.orbital_period * 60.0),
            self.orbit_normal,
        )) - vector_lf
    }

    pub fn get_estimate_point_offset(&self, eta: f64) -> f32 {
        let num: f64 = PI * eta * 360.0 / self.orbital_period;
        self.orbit_radius * 40000.0 * 0.5 * (num.powi(2) as f32)
    }
}
